src/main/kotlin/org/rust/ide/annotator/RustHighlightingAnnotator.kt

```diff
@@ -12,10 +11,40 @@ import org.rust.lang.core.psi.util.elementType
 
 // Highlighting logic here should be kept in sync with tags in RustColorSettingsPage
 class RustHighlightingAnnotator : Annotator {
-    override fun annotate(element: PsiElement, holder: AnnotationHolder) = element.accept(object : RustElementVisitor() {
-        override fun visitAttr(o: RustAttrElement) {
-            holder.highlight(o, RustColor.ATTRIBUTE)
+
+    override fun annotate(element: PsiElement, _holder: AnnotationHolder) {
+        val highlighter = wrap(_holder)
+        if (element is RustReferenceElement) {
+            // Highlight the element dependent on what it's referencing.
+            val ref = element.reference.resolve() ?: return
+
+            val text = when (element) {
+                is RustPathElement -> element.identifier
+                else -> element
+            }
+            resolveColor(ref)?.let { highlighter.highlight(text, it) }
+        } else {
+            element.accept(highlightingVisitor(highlighter))
         }
+    }
+
+    // Capture the color of the element w/o mutating anything.
+    fun resolveColor(e: PsiElement): RustColor? {
+        var outColor: RustColor? = null;
+        val resolver = highlightingVisitor(object : Highlighter { // @TODO Could use RustComputingVisitor
```

Comment:
Please do, I hope that data class for highlighting info would be much clearer then `Highlighter` interface.